Add onBatchStarted callback to "BatchCompleteListener".

kmagiera opened the pull request:
This change adds a way for native modules to receive "batchStarted" event (I've actually renamed OnBatchCompleteListener to JSBatchListener so that the name does no longer suggest it can be used only for getting "batchComplete" events).

I think this callback will be important for implementing features that may want to interact with native modules from inside of the native code and will allow to isolate modifications of UI that are done by JS from the ones that can be driven by purely native code. A good example would be driving animations from native code. In that case we'd like for the animation code to be able to update view properties (such as transforms) but also props that affect shadow view hierarchy (layout-related props). Since changes made to layout props have to go though the regular prop update pipeline (update shadow view -> calculate layout on batch complete -> update values in UI thread) the best way IMO is to "trigger" onBatchComplete when animation frame is requested and let UIManager start all the view updating logic. The issue is with other updates that might be coming from JS at the same time. SO the solution is to detect if JS is in a process of sending native commands and wait until the process is finished - that's where this new callback I'm adding will help a lot.

Test plan:
Add boolean "batchState" variable in UImanagerModule. Update it to tru in #onBatchStarted and to false in #onBatchComplete. Add asserts in #onBatchComplete for "batchState" to be true when entering the method -> play with some apps, test reload, asserts should always be passing

bestander commented:
@kmagiera FYI build has failed https://circleci.com/gh/facebook/react-native/1866

brentvatne commented:
@bestander - seems like build is failing on other PRs too

bestander commented:
@brentvatne yeah, thanks.
There was an accident break today but it got fixed just 15 minutes ago.

kmagiera commented:
@bestander I've updated my PR yesterday with an import statement that has been removed after I rebased but still getting failures on circleCI. Can you please check if those are related to my commit as while running them locally all unit test pass for me and "connectedAndroidTest" fail with and without my change on HorizontalScrollViewTest (which is sth different that what circleCI reports).

bestander commented:
Thanks, @kmagiera.
Sometimes the test runner can be a bit flaky on a CI VM, I'll have a look if that is the case.

bestander commented:
Yeah, looks like a random issue, thanks for keeping an eye on it.
I'll add a limit to memory and number of threads for buck

astreet commented:
noooooobee

We're trying to get rid of the concept of onBatchComplete as well in favor of an explicit UIManager#flush() operation. Can we have native animations integrate more tightly with the UIManager module instead of extending this implicit bridge contract?

kmagiera commented:
@astreet Sure, this just seemed to be the most elegant solution. As I described in the comment when I schedule updates from native animated "thread" I also need to "flush" the batch. So what is required from native animated perspective is the knowledge of UI manager transaction boundaries. If as a part of migrating towards "UIManager#flush()" we can also add "UIManager#startUITransaction" that would be perfect for me. Otherwise the alternative is to add a piece of code in every single UIManager method that enqueues ui operations

astreet commented:
First, I want to point out an issue with off-JS thread animations that go through the shadow hierarchy: if we want to use the same mechanism to enable off-JS thread gestures, we can't easily go through the shadow hierarchy for some things, the main one being gestures tied to scroll position, e.g. sticky headers. This is because the scroll position would always be a frame ahead of the actual sticky header position because shadow thread updates that are posted on frame N have layout calculated and are then posted to the main thread for display on frame N+1. For other things, maybe this won't be noticeable. However, I'm concerned about supporting both the set of attributes that can be applies on frame N (translate/scale/rotation) and those that can be applied on frame N+1 (CSS layout updates). Generally, it's expensive to recalculate layout on every frame of an animation anyway. To make sure, do you have cases where you need these 'frame N+1' animations?

As far as implementing this with flush, an explicit batch-start call from JS could work, though I'm not sure all UIManager calls go through the same path in JS. What are your thoughts on just determining this in the native module itself (e.g. having createView etc set a flag that is cleared on flush)?

kmagiera commented:
@astreet I built my prototype with NavigationRFC implementation in mind, which was using css-layout props to animate the views back then (not sure what the current status is, as there was a fix for changing this to translate animation but I think it got reverted).

I agree that when only non-css props are being updated as a part of animation we shouldn't go through the "native thread". And I think the first version I'll send out will only support that type of animations, so I'm not going to urge on this PR. I also proposed "adding a flag" to each UIManager method in my previous comment but IMO it would be a very error-prone approach and I'd prefer an explicit call in order to start UI transaction.

astreet commented:
Ok, I'm fine with an explicit 'batch start' command.

kmagiera commented:
Closing this for now. I decided to go with an approach that doesn't allow for the layout properties to be animated in native as a first iteration. This is no longer necessary and, perhaps, this change is already outdated
